[front] - feat(obs): treemap search

JulesBelveze opened the pull request:
## Description

This PR adds a treemap visualization to the agent observability dashboard to show document retrieval patterns by datasource. The chart aggregates Elasticsearch data showing how many documents are retrieved per search, grouped by datasource and organized by MCP server configuration. The feature is gated behind the `agent_tool_outputs_analytics` feature flag and builds on the analytics infrastructure introduced in PR #19311.

The treemap displays user-friendly names for managed datasources by mapping internal naming conventions to connector provider names, and uses custom names for MCP server configurations when available.

## Risk

Low - feature is behind the `agent_tool_outputs_analytics` feature flag.

## Tests

- [x] Locally
- [x] Front-edge

## Deploy Plan

Deploy front

JulesBelveze committed:
[front/pages/api/observability] - feature: add API for datasource retrieval metrics

 - Implement a new endpoint to fetch metrics associated with datasource retrieval for an agent
 - Allow query by days and versions with defaults set for unspecified queries
 - Include authentication and permissions checks to ensure security and proper access control

JulesBelveze committed:
[assistant/observability] - feature: implement datasource retrieval metrics fetching

 - Add new function to query Elasticsearch for datasource retrieval metrics by MCP server configs
 - Aggregate results to include MCP server name and count of documents per datasource

JulesBelveze committed:
[swr] - feature: add hook for agent datasource retrieval metrics

 - Implement a new SWR hook to fetch and provide observability metrics related to datasource retrieval within agents
 - The hook includes data like total retrievals and loading/error states handling for UI updates based on the fetched data

JulesBelveze committed:
[front/components/agent_builder/observability/charts] - feature: add treemap chart for datasource retrieval

 - Implement a new chart component to visualize document retrieval counts per MCP server organized by datasource
 - Utilize recharts library to create a responsive Treemap with custom styling and tooltips for enhanced data representation
 - Integrate the component with the observability context to display data based on selected time period and version filter

JulesBelveze committed:
[front/components/agent_builder/observability/charts] - refactor: enhance TreemapNode interface and remove debug logs

 - Enhanced the `TreemapNode` interface to allow dynamic keys to support future custom properties
 - Removed unnecessary console log calls for cleaner and more efficient code execution

[front/components/observability] - feature: integrate DatasourceRetrievalTreemapChart component

 - Added `DatasourceRetrievalTreemapChart` component to the `AgentObservability` to visualize data sources retrieval performance
 - Implemented visual separation with a `Separator` for better layout structure and readability

JulesBelveze committed:
[front/observability] - feature: conditionally display DatasourceRetrievalTreemapChart

 - Integrate 'agent_tool_outputs_analytics' feature flag to toggle visibility of the DatasourceRetrievalTreemapChart
 - Ensure the feature is only accessible to users when the feature flag is enabled in their workspace

JulesBelveze committed:
[agent_builder/observability] - feature: display friendly names for managed data sources in treemap chart

 - Translate internal data source names into user-friendly names for better readability in the observability dashboard
 - Add utility function to map from internal data source naming conventions to display names using connector configurations

JulesBelveze committed:
[front/components/agent_builder/observability/charts] - refactor: update chart titles and descriptions for clarity

- Changed the title to better reflect the chart's purpose as it moves into beta
- Updated the description to accurately describe the data being displayed as retrieved per search, grouped by datasource

JulesBelveze committed:
[agent_builder/observability/charts] - refactor: update treemap chart component props and structure

 - Renamed 'CustomizedContent' to 'TreemapContent' for better clarity in Treemap chart component
 - Added 'DatasourceRetrievalTreemapChartProps' for type safety and to clarify required props
 - Replaced generic 'workspaceId' with 'owner' object to have more specific type 'WorkspaceType'
 - Used 'isMCPServerConfiguration' type guard for better type discrimination in action filtering
 - Renamed prop passed to 'TreemapContent' from '<CustomizedContent />' to '<TreemapContent />'

JulesBelveze committed:
[agent_builder/observability/charts] - refactor: standardized workspaceId prop in treemap components

- Replaced `owner` prop with `workspaceId` to ensure consistency across treemap components
- Updated corresponding hook calls to use the new `workspaceId` prop for fetching data

JulesBelveze committed:
[front] - refactor: display action names using display formatter in treemap

- Introduced use of `asDisplayName` function to format action names for treemap nodes
- Ensured consistency in displaying connector provider names across the application

flvndvd commented on front/components/agent_builder/observability/charts/DatasourceRetrievalTreemapChart.tsx:
```diff
@@ -0,0 +1,253 @@
+            textAnchor="middle"
+            className="pointer-events-none fill-foreground text-xs font-medium dark:fill-foreground-night"
+          >
+            {name.length > 20 ? `${name.slice(0, 17)}...` : name}
```
IIRC Tailwind has a class for this.

flvndvd commented on front/lib/api/assistant/observability/datasource_retrieval.ts:
```diff
@@ -0,0 +1,115 @@
+  by_mcp_server_config?: estypes.AggregationsMultiBucketAggregateBase<McpServerConfigBucket>;
+};
+
+export async function fetchDatasourceRetrievalMetrics(
```
To make things safer, I'd at least pass `auth` here and ensure the `workspaceId` clause is set in this function, closer to the call site.

flvndvd commented on front/pages/api/w/[wId]/assistant/agent_configurations/[aId]/observability/datasource-retrieval.ts:
```diff
@@ -0,0 +1,121 @@
+          status_code: 400,
+          api_error: {
+            type: "invalid_request_error",
+            message: `Invalid query parameters: ${q.error.message}`,
```
You should use `fromError(q.error).toString()`

JulesBelveze committed:
[api/observability] - refactor: standardize error handling for datasource retrieval

 - Use zod-validation-error to convert Zod errors into a standardized format
 - Implement isString utility function for type-checking instead of direct typeof comparison
 - Set default value for 'days' query parameter using Zod's schema defaults

JulesBelveze committed:
[assistant/observability] - feature: enhance datasource retrieval metrics query

 - Refactor fetchDatasourceRetrievalMetrics function to accept an object with workspaceId, agentId, days, and version
 - Introduce buildAgentAnalyticsBaseQuery utility for better query construction based on new parameters
 - Update useAgentDatasourceRetrieval SWR hook to construct the API query with URLSearchParams for better parameter handling

JulesBelveze committed:
[agent_builder/observability/charts] - fix: improve DatasourceRetrievalTreemapChart text display

- Truncate and add an ellipsis to text overflow within treemap nodes for readability
- Initialize name map for MCP configurations more efficiently with early return

JulesBelveze commented on front/components/agent_builder/observability/charts/DatasourceRetrievalTreemapChart.tsx:
```diff
@@ -0,0 +1,253 @@
+          </text>
+          <text
+            x={x + width / 2}
+            y={y + height / 2 + 8}
```
https://github.com/dust-tt/dust/pull/19496#discussion_r2661443805

JulesBelveze committed:
[front/lib/api/assistant] - refactor: streamline datasource retrieval metrics fetch

 - Refactor `fetchDatasourceRetrievalMetrics` to accept `Authenticator` and encapsulate workspace retrieval
 - Update the corresponding API handler to pass `auth` and simplified parameters to the refactored function

JulesBelveze committed:
[agent_builder/observability/charts] - refactor: abstract label offset in treemap chart

- Extract hard-coded label offset value into a constant for clarity and maintainability
- Adjust usage of the label offset in the `TreemapContent` component for consistent label positioning

JulesBelveze committed:
[agent_builder/observability] - refactor: streamline datasource retrieval data fetching

 - Removed dependency on agent configuration for MCP server display names
 - Enhanced DatasourceRetrievalTreemapChart to use server config name directly from retrieval data
 - Simplified fetch logic by using AgentMCPServerConfigurationResource for server config names
 - Reduced complexity in treemap chart legend creation by removing mcpConfigNameMap usage

JulesBelveze committed:
[front/components/agent_builder/observability/charts] - refactor: improve text visibility logic and rendering in treemap chart

 - Changed visibility condition for text labels to display for any non-zero-sized rectangle
 - Replaced individual SVG text elements with a foreignObject and divs to handle text overflow and positioning

JulesBelveze committed:
[agent_builder/observability] - refactor: streamline display name retrieval for data sources

 - Remove local getDisplayNameForDataSourceName function in favor of centralized getDisplayNameForDataSource
 - Use dataSourceId and displayName directly in observability charts for more consistency and maintainability

JulesBelveze committed:
[front/lib/api/assistant/observability] - fix: correct dataSource map key and exclude JIT MCP servers

 - Change the key for the dataSource map from 'sId' to 'Id' to ensure consistency
 - Add a filter to exclude Just-In-Time (JIT) MCP servers from the retrieval metrics processing
